tests/functional/s3/test_presign_command.py

```diff
@@ -0,0 +1,167 @@
+
+        This method compares an actual URL against a dict of expected
+        values.  The reason that the "expected_match" is a dict instead
+        of the expected presigned URL is because the query params
+        are unordered so we can't guarantee an expected query param
+        ordering.
+
+        """
+        parts = urlsplit(actual_url)
+        self.assertEqual(parts.netloc, expected_match['hostname'])
+        self.assertEqual(parts.path, expected_match['path'])
+        query_params = self.parse_query_string(parts.query)
+        self.assertEqual(query_params, expected_match['query_params'])
+
+    def parse_query_string(self, query_string):
+        pairs = []
+        for part in query_string.split('&'):
+            pairs.append(part.split('=', 1))
+        return dict(pairs)
+
+    def get_presigned_url_for_cmd(self, cmdline):
+        with mock.patch('time.time', FROZEN_TIME):
+            with mock.patch('datetime.datetime') as d:
+                d.utcnow = FROZEN_DATETIME
+                stdout = self.assert_params_for_cmd(cmdline, None)[0].strip()
+                return stdout
+
+    def test_generates_a_url(self):
+        stdout = self.get_presigned_url_for_cmd(
+            self.prefix + 's3://bucket/key')
+
+        self.assert_presigned_url_matches(
+            stdout, {
+                'hostname': 'bucket.s3.amazonaws.com',
+                'path': '/key',
+                'query_params': {
+                    'AWSAccessKeyId': 'access_key',
+                    'Expires': str(FROZEN_TIMESTAMP + DEFAULT_EXPIRES),
+                    'Signature': '2m9M0eLB%2BqI0nUpkyTskKmHd0Ig%3D',
+                }
+            }
+        )
+
+    def test_handles_non_dns_compatible_buckets(self):
+        stdout = self.get_presigned_url_for_cmd(
+            self.prefix + 's3://bucket.dots/key')
+
+        self.assert_presigned_url_matches(
+            stdout, {
+                'hostname': 's3.amazonaws.com',
+                'path': '/bucket.dots/key',
+                'query_params': {
+                    'AWSAccessKeyId': 'access_key',
+                    'Expires': str(FROZEN_TIMESTAMP + DEFAULT_EXPIRES),
+                    'Signature': '0IiC2vxub438EVcKfEFEMHuoHRw%3D',
+                }
+            }
+        )
+
+    def test_handles_expires_in(self):
+        expires_in = 1000
+        stdout = self.get_presigned_url_for_cmd(
+            self.prefix + 's3://bucket/key --expires-in %s' % expires_in)
+
+        self.assert_presigned_url_matches(
+            stdout, {
+                'hostname': 'bucket.s3.amazonaws.com',
+                'path': '/key',
+                'query_params': {
+                    'AWSAccessKeyId': 'access_key',
+                    'Expires': str(FROZEN_TIMESTAMP + expires_in),
+                    'Signature': 'WZEMcfBNlzfTZBq3bOvYef1cfoU%3D',
+                }
+            }
+        )
+
+    def test_handles_sigv4(self):
```

Comment:
S3 accelerate doesn't actually work because of https://github.com/boto/botocore/issues/978.  We don't emit the request-created event so we never switch out the hosts.

I can add tests for the addressing styles though.